cmakelib/lexer/filter.go

```diff
@@ -0,0 +1,161 @@
+/*
+ * Copyright 2019 The Kythe Authors. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package lexer
+
+import (
+	"github.com/alecthomas/participle/lexer"
+)
+
+// filterLexer processes the raw stream of tokens coming from the underlying scanner-based lexer
+// to make the resulting AST simpler and more readable.
+type filterLexer struct {
+	l    lexer.Lexer
+	buf  []lexer.Token
+	prev lexer.Token
+}
+
+// Next implements the lexer.Lexer interface for filterLexer.
+func (l *filterLexer) Next() (lexer.Token, error) {
+	var err error
+	if len(l.buf) > 0 {
+		tok := l.buf[0]
+		l.buf = l.buf[1:]
+		return tok, nil
+	}
+	switch l.prev.Type {
+	case BracketOpen:
+		return l.bufferTokens(combineBracketContent(l.l, len(l.prev.Value)))
+	case Quote:
+		return l.bufferTokens(combineQuotedContent(l.l))
+	}
+	for {
+		l.prev, err = l.l.Next()
+		if err != nil {
+			break
+		}
+		switch l.prev.Type {
+		case Comment:
+			l.prev, err = l.bufferTokens(consumeComment(l.l))
+			if err != nil || l.prev.Type == lexer.EOF {
+				goto done
+			}
+		default:
+			goto done
+		}
+	}
+done:
```

Comment:
This is probably fine, but I suggest instead labelling the `for` and breaking from that, e.g.

```go
scan:
  for {
     …
     switch l.prev.Type {
     case Comment:
         break scan
   }
```

just so that the reader doesn't have to read forward to resolve the jump target. It doesn't change the meaning, though.

Reply:
That was what I'd gone for originally, but didn't realize the label needed to be before the loop, thanks!